src/wasm/wasm-validator.cpp

```diff
@@ -2898,9 +2913,18 @@ static void validateTables(Module& module, ValidationInfo& info) {
                         segment->offset,
                         "table segment offset should be reasonable");
     }
-    for (auto name : segment->data) {
+    // Avoid double checking items
+    if (module.features.hasReferenceTypes()) {
       info.shouldBeTrue(
-        module.getFunctionOrNull(name), name, "segment name should be valid");
+        std::all_of(segment->data.begin(),
+                    segment->data.end(),
+                    [](Expression* expr) {
+                      return expr->is<RefFunc>() ||
+                             (expr->is<RefNull>() &&
+                              Type::isSubType(expr->type, Type::funcref));
+                    }),
+        "elem",
+        "element segment items must be either ref.func or ref.null func");
     }
   }
 }
```

Comment:
Thanks for this, I was under the impression that the module will be walked during validation so expressions will be visited automatically. Now, I tried to use `FunctionValidator`, but looks like `ref.*` validation rules refuse to validate the module if reference-types is not enabled. I can validate existence of functions separately and not use `FunctionValidator` for now, until we come up with a better a way maybe?

Reply:
So I extracted the main validation logic from `visitRefFunc` and `visitRefNull` into `validateRefFunc` and `validateRefNull` to avoid duplication, and used those in `validateTables`. It works, but looks hacky.

Reply:
> looks like ref.* validation rules refuse to validate the module if reference-types is not enabled.

That is the bigger issue here, yes. This PR uses a reference type thing even in cases where reference types are not enabled. There is one example of that so far, using more specific types for `ref.func` even if typed function references are not enabled, but that one was carefully seen to be safe. This one looks like it would take a little more work.

But maybe not much more. Can we just check "if !getFunction then it's ok to use refFunc`? That is, allow it in global locations, but not in function bodies.

Reply:
That's what I was looking for! Thanks a lot for the direction. :) It's fixed now.